File: src/signals/dispatcher.js
// Cable-driven dispatcher.
//
// One graph: cables (within a room) plus the topology bridge (a terminal in
// room R can be peered to a terminal in room R'; that's external state the
// caller owns). The dispatcher walks the graph from an emitting jack and
// delivers signals to receiving jacks via per-processor input handlers.
//
// Hop cap (default 32): every traversal hop increments hopCount; when the cap
// is reached the signal is dropped. Cycles end silently. (Routing the cycle
// diagnostic upward to the nearest S5 algedonic channel is a future concern;
// today we just stop.)
//
// Dedup: each processor keeps a small bounded set of recently-seen signal ids
// so the same signal arriving via two paths only fires the input handler once.
//
// Cross-room: when a jack→terminal cable is followed, the dispatcher calls
// `onTerminal(fromRoomKey, terminalId, signal, hopCount)`. The caller
// resolves topology and (if there's a peer) calls
// `deliverFromTerminal(peerRoomKey, peerTerminalId, signal, hopCount)`. The
// dispatcher then continues delivery from terminal-sourced cables in the
// peer room.
//
// Broadcast: per-instance flag. When set, an emit fans out to every terminal
// in the emitter's room (caller registers `roomTerminalsByRoom`). Internal
// jack→jack cables continue to fire. External jack→terminal cables are
// suppressed (broadcast wins for terminal egress).

const DEFAULT_HOP_CAP = 32
const DEFAULT_DEDUP_SIZE = 256

export function createDispatcher({
  onTerminal,
  hopCap = DEFAULT_HOP_CAP,
  dedupSize = DEFAULT_DEDUP_SIZE,
} = {}) {
  // ---- mutable state -----------------------------------------------------
  let cablesByRoom = {}
  let roomTerminalsByRoom = {}
  const broadcastByInstance = {}
  const inputHandlers = {}        // instanceId → handler({ signal, portId })
  const instanceLocation = {}     // instanceId → roomKey
  const seenByInstance = {}       // instanceId → Set<signalId> (bounded)

  // ---- mutators ----------------------------------------------------------
  function setCables(next) { cablesByRoom = next || {} }
  function setRoomTerminals(next) { roomTerminalsByRoom = next || {} }
  function setBroadcast(instanceId, broadcast) {
    if (broadcast) broadcastByInstance[instanceId] = true
    else delete broadcastByInstance[instanceId]
  }
  function registerProcessor(instanceId, { roomKey, inputHandler }) {
    instanceLocation[instanceId] = roomKey
    inputHandlers[instanceId] = inputHandler
    if (!seenByInstance[instanceId]) seenByInstance[instanceId] = new Set()
  }
  function unregisterProcessor(instanceId) {
    delete instanceLocation[instanceId]
    delete inputHandlers[instanceId]
    delete seenByInstance[instanceId]
    delete broadcastByInstance[instanceId]
  }

  // ---- internal: deliver to a single jack target -------------------------
  function deliverToJack(signal, target, hopCount) {
    if (hopCount > hopCap) return
    const handler = inputHandlers[target.instanceId]
    if (!handler) return

    // Dedup by signal id at this processor
    const seen = seenByInstance[target.instanceId]
    if (seen && signal.id) {
      if (seen.has(signal.id)) return
      seen.add(signal.id)
      if (seen.size > dedupSize) {
        const it = seen.values()
        seen.delete(it.next().value) // drop oldest
      }
    }

    handler({ signal: { ...signal, hopCount }, portId: target.portId })
  }

  // ---- emit: jack → graph ------------------------------------------------
  function emit(signal, { fromInstanceId, fromPortId }) {
    const startHop = signal.hopCount || 0
    if (startHop >= hopCap) return
    const startRoomKey = instanceLocation[fromInstanceId]
    if (!startRoomKey) return

    const broadcasting = !!broadcastByInstance[fromInstanceId]
    const cables = cablesByRoom[startRoomKey] || []
    const fanoutHop = startHop + 1
    const sentToTerminal = new Set()

    for (const cable of cables) {
      const s = cable.source
      if (s?.kind !== 'jack') continue
      if (s.instanceId !== fromInstanceId) continue
      if (s.portId !== fromPortId) continue

      const t = cable.target
      if (t?.kind === 'jack') {
        // Internal cables always fire — broadcast does not suppress them.
        deliverToJack(signal, t, fanoutHop)
      } else if (t?.kind === 'terminal' && !broadcasting) {
        // External cables only fire when not broadcasting.
        if (!sentToTerminal.has(t.terminalId)) {
          sentToTerminal.add(t.terminalId)
          onTerminal?.(startRoomKey, t.terminalId, signal, fanoutHop)
        }
      }
    }

    // Broadcast: fan out to every terminal in the emitter's room.
    if (broadcasting) {
      const tids = roomTerminalsByRoom[startRoomKey] || []
      for (const tid of tids) {
        if (sentToTerminal.has(tid)) continue
        sentToTerminal.add(tid)
        onTerminal?.(startRoomKey, tid, signal, fanoutHop)
      }
    }
  }

  // ---- arrival from peer room: terminal → graph --------------------------
  function deliverFromTerminal(roomKey, terminalId, signal, hopCount = 0) {
    if (hopCount > hopCap) return
    const cables = cablesByRoom[roomKey] || []
    const fanoutHop = hopCount + 1
    const sentToTerminal = new Set([terminalId]) // don't bounce back

    for (const cable of cables) {
      const s = cable.source
      if (s?.kind !== 'terminal' || s.terminalId !== terminalId) continue

      const t = cable.target
      if (t?.kind === 'jack') {
        deliverToJack(signal, t, fanoutHop)
      } else if (t?.kind === 'terminal') {
        if (!sentToTerminal.has(t.terminalId)) {
          sentToTerminal.add(t.terminalId)
          onTerminal?.(roomKey, t.terminalId, signal, fanoutHop)
        }
      }
    }
  }

  return {
    setCables, setRoomTerminals, setBroadcast,
    registerProcessor, unregisterProcessor,
    emit, deliverFromTerminal,
    // Test helpers — read-only views of internal state
    _state: () => ({
      cablesByRoom, roomTerminalsByRoom,
      broadcastByInstance: { ...broadcastByInstance },
      registeredInstances: Object.keys(inputHandlers),
    }),
  }
}

File: server/dispatcher.go
package main

import (
	"sync"
)

const (
	MaxHops     = 32
	DedupeLimit = 256
)

// Dispatcher routes signals through the cable graph
type Dispatcher struct {
	mu sync.RWMutex

	// Cable graph per room
	cablesByRoom map[string][]Cable

	// Registered processors
	processors map[string]*ProcessorRegistration

	// Dedup: recently seen signal IDs per processor (LRU-ish, bounded)
	seenByInstance map[string][]string

	// Broadcast mode per instance
	broadcastByInstance map[string]bool

	// Terminal IDs per room (for broadcast fan-out)
	roomTerminals map[string][]string

	// Callback for cross-room delivery
	onTerminal func(fromRoomKey, terminalID string, signal *Signal, hopCount int)
}

// ProcessorRegistration holds a registered processor's info
type ProcessorRegistration struct {
	RoomKey      string
	InputHandler func(signal *Signal, portID string)
}

// NewDispatcher creates a new dispatcher
func NewDispatcher() *Dispatcher {
	return &Dispatcher{
		cablesByRoom:        make(map[string][]Cable),
		processors:          make(map[string]*ProcessorRegistration),
		seenByInstance:      make(map[string][]string),
		broadcastByInstance: make(map[string]bool),
		roomTerminals:       make(map[string][]string),
	}
}

// SetOnTerminal sets the callback for cross-room delivery
func (d *Dispatcher) SetOnTerminal(fn func(fromRoomKey, terminalID string, signal *Signal, hopCount int)) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.onTerminal = fn
}

// SetCables updates the cable graph
func (d *Dispatcher) SetCables(cables map[string][]Cable) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.cablesByRoom = cables
}

// SetRoomTerminals sets terminal IDs per room
func (d *Dispatcher) SetRoomTerminals(terminals map[string][]string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.roomTerminals = terminals
}

// RegisterProcessor registers a processor for receiving signals
func (d *Dispatcher) RegisterProcessor(instanceID string, roomKey string, inputHandler func(signal *Signal, portID string)) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.processors[instanceID] = &ProcessorRegistration{
		RoomKey:      roomKey,
		InputHandler: inputHandler,
	}
	d.seenByInstance[instanceID] = []string{}
}

// UnregisterProcessor removes a processor registration
func (d *Dispatcher) UnregisterProcessor(instanceID string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	delete(d.processors, instanceID)
	delete(d.seenByInstance, instanceID)
	delete(d.broadcastByInstance, instanceID)
}

// SetBroadcast sets broadcast mode for a processor
func (d *Dispatcher) SetBroadcast(instanceID string, broadcast bool) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.broadcastByInstance[instanceID] = broadcast
}

// Emit sends a signal from a processor's output port
func (d *Dispatcher) Emit(signal *Signal, fromInstanceID, fromPortID string, hopCount int) {
	if hopCount > MaxHops {
		return // Cycle protection
	}

	d.mu.RLock()
	reg := d.processors[fromInstanceID]
	if reg == nil {
		d.mu.RUnlock()
		return
	}
	roomKey := reg.RoomKey
	cables := d.cablesByRoom[roomKey]
	broadcast := d.broadcastByInstance[fromInstanceID]
	terminals := d.roomTerminals[roomKey]
	onTerminal := d.onTerminal
	d.mu.RUnlock()

	// If broadcast mode, send to all terminals
	if broadcast {
		for _, terminalID := range terminals {
			if onTerminal != nil {
				cloned := signal.Clone()
				onTerminal(roomKey, terminalID, cloned, hopCount+1)
			}
		}
		return
	}

	// Find cables from this port
	for _, cable := range cables {
		if cable.Source.Kind != "jack" {
			continue
		}
		if cable.Source.InstanceID != fromInstanceID || cable.Source.PortID != fromPortID {
			continue
		}

		// Deliver based on target type
		if cable.Target.Kind == "jack" {
			// Internal delivery to another processor
			d.deliverToProcessor(signal, cable.Target.InstanceID, cable.Target.PortID, hopCount+1)
		} else if cable.Target.Kind == "terminal" {
			// Cross-room delivery
			if onTerminal != nil {
				cloned := signal.Clone()
				onTerminal(roomKey, cable.Target.TerminalID, cloned, hopCount+1)
			}
		}
	}
}

// DeliverFromTerminal handles signals arriving from another room
func (d *Dispatcher) DeliverFromTerminal(roomKey, terminalID string, signal *Signal, hopCount int) {
	if hopCount > MaxHops {
		return
	}

	d.mu.RLock()
	cables := d.cablesByRoom[roomKey]
	d.mu.RUnlock()

	// Find cables from this terminal
	for _, cable := range cables {
		if cable.Source.Kind != "terminal" || cable.Source.TerminalID != terminalID {
			continue
		}

		if cable.Target.Kind == "jack" {
			d.deliverToProcessor(signal, cable.Target.InstanceID, cable.Target.PortID, hopCount+1)
		}
		// terminal→terminal cables don't make sense within a room
	}
}

// deliverToProcessor sends a signal to a processor's input
func (d *Dispatcher) deliverToProcessor(signal *Signal, instanceID, portID string, hopCount int) {
	d.mu.RLock()
	reg := d.processors[instanceID]
	seen := d.seenByInstance[instanceID]
	d.mu.RUnlock()

	if reg == nil {
		return
	}

	// Dedup check
	for _, id := range seen {
		if id == signal.ID {
			return // Already seen this signal
		}
	}

	// Add to seen list (bounded)
	d.mu.Lock()
	d.seenByInstance[instanceID] = append(d.seenByInstance[instanceID], signal.ID)
	if len(d.seenByInstance[instanceID]) > DedupeLimit {
		d.seenByInstance[instanceID] = d.seenByInstance[instanceID][1:]
	}
	d.mu.Unlock()

	// Call the input handler
	if reg.InputHandler != nil {
		reg.InputHandler(signal, portID)
	}
}
